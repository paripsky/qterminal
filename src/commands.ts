type CommandContext = {
  history: string[];
  clear: () => void;
};

type Command = {
  description: string;
  handler: (args: string[], context: CommandContext) => string;
};

function tryEval(command: string) {
  try {
    return eval(command);
  } catch (e) {
    return (e as Error).message;
  }
}

const commands: Record<string, Command> = {
  help: {
    description: 'Show help',
    handler: () => {
      return Object.keys(commands).reduce(
        (acc, command) => `${acc}${command}: ${commands[command].description}\n`,
        'Available Commands:\n',
      );
    },
  },
  smiley: {
    description: 'A smiley face',
    handler: () => ':)\n:)',
  },
  history: {
    description: 'Show command history',
    handler: (_, { history }) => history.join('\n'),
  },
  eval: {
    description: 'Evaluate a JavaScript expression',
    handler: (args: string[]) => tryEval(args.join(' ')).toString(),
  },
  clear: {
    description: 'Clear the output',
    handler: (_, { clear }) => {
      clear();
      return 'Cleared!';
    },
  },
};

export function handleCommand(command: string, context: CommandContext) {
  const cmd = command.split(' ')[0].toLowerCase();
  const args = command.split(' ').slice(1);

  if (!commands[cmd]) return `${cmd}: command not found`;

  return commands[cmd].handler(args, context);
}
