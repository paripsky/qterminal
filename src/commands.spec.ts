import { describe, expect, it, vi } from 'vitest';

import { handleCommand } from './commands';

const mockAPI = {
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  clear: () => {},
  history: [],
};

describe('commands', async () => {
  it('clear', () => {
    const clear = vi.fn();
    const result = handleCommand('clear', { history: [], clear });
    expect(result).toBe('Cleared!');
    expect(clear).toHaveBeenCalled();
  });

  it('history', () => {
    const result = handleCommand('history', { ...mockAPI, history: ['foo', 'bar'] });
    expect(result).toBe('foo\nbar');
  });

  it('help', () => {
    const result = handleCommand('help', { ...mockAPI });
    expect(result).toBe(`Available Commands:
help: Show help
smiley: A smiley face
history: Show command history
eval: Evaluate a JavaScript expression
clear: Clear the output
`);
  });
});
