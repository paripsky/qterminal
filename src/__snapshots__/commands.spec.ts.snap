// Vitest Snapshot v1

exports[`html creates dom elements > returns only the first child 1`] = `
<div>
  test div creation
</div>
`;

exports[`html creates dom elements > should match snapshot 1`] = `
<div>
  test div creation
</div>
`;
